File: 02. Two Pointers and Sliding Window/03. Longest Subtring without repeating characters/main.cpp
#include <iostream>
#include <unordered_set>
#include <unordered_map>
using namespace std;

// 1. Brute Force Solution - O(n^3)
bool allUnique(string s, int start, int end) {
    unordered_set<char> chars;
    for (int i = start; i <= end; i++) {
        if (chars.find(s[i]) != chars.end())
            return false;
        chars.insert(s[i]);
    }
    return true;
}

int bruteForce(string s) {
    int n = s.length(), maxLength = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i; j < n; j++) {
            if (allUnique(s, i, j)) {
                maxLength = max(maxLength, j - i + 1);
            }
        }
    }
    return maxLength;
}

// 2. Improved Brute Force using Hash Set - O(n^2)
int improvedBruteForce(string s) {
    int n = s.length(), maxLength = 0;
    for (int i = 0; i < n; i++) {
        unordered_set<char> seen;
        for (int j = i; j < n; j++) {
            if (seen.find(s[j]) != seen.end()) break;
            seen.insert(s[j]);
            maxLength = max(maxLength, j - i + 1);
        }
    }
    return maxLength;
}

// 3. Optimal Sliding Window using Hash Map - O(n)
int slidingWindow(string s) {
    unordered_map<char, int> lastSeen;
    int left = 0, maxLength = 0;
    
    for (int right = 0; right < s.length(); right++) {
        if (lastSeen.find(s[right]) != lastSeen.end()) {
            left = max(left, lastSeen[s[right]] + 1);
        }
        lastSeen[s[right]] = right;
        maxLength = max(maxLength, right - left + 1);
    }
    
    return maxLength;
}

// 4. Sliding Window using Hash Set - O(n)
int slidingWindowSet(string s) {
    unordered_set<char> seen;
    int left = 0, maxLength = 0;
    
    for (int right = 0; right < s.length(); right++) {
        while (seen.find(s[right]) != seen.end()) {
            seen.erase(s[left]);
            left++;
        }
        seen.insert(s[right]);
        maxLength = max(maxLength, right - left + 1);
    }
    
    return maxLength;
}

int main() {
    string s;
    cout << "Enter a string: ";
    cin >> s;

    cout << "Brute Force: " << bruteForce(s) << endl;
    cout << "Improved Brute Force: " << improvedBruteForce(s) << endl;
    cout << "Sliding Window (Hash Map): " << slidingWindow(s) << endl;
    cout << "Sliding Window (Hash Set): " << slidingWindowSet(s) << endl;

    return 0;
}

File: 02. Two Pointers and Sliding Window/03. Longest Subtring without repeating characters/main.py
def brute_force(s):
    def all_unique(sub):
        return len(set(sub)) == len(sub)

    max_length = 0
    for i in range(len(s)):
        for j in range(i, len(s)):
            if all_unique(s[i:j+1]):
                max_length = max(max_length, j - i + 1)
    return max_length

def improved_brute_force(s):
    max_length = 0
    for i in range(len(s)):
        seen = set()
        for j in range(i, len(s)):
            if s[j] in seen:
                break
            seen.add(s[j])
            max_length = max(max_length, j - i + 1)
    return max_length

def sliding_window(s):
    last_seen = {}
    left = max_length = 0
    for right, char in enumerate(s):
        if char in last_seen:
            left = max(left, last_seen[char] + 1)
        last_seen[char] = right
        max_length = max(max_length, right - left + 1)
    return max_length

def sliding_window_set(s):
    seen = set()
    left = max_length = 0
    for right in range(len(s)):
        while s[right] in seen:
            seen.remove(s[left])
            left += 1
        seen.add(s[right])
        max_length = max(max_length, right - left + 1)
    return max_length

def main():
    s = input("Enter a string: ")
    print("Brute Force:", brute_force(s))
    print("Improved Brute Force:", improved_brute_force(s))
    print("Sliding Window (Hash Map):", sliding_window(s))
    print("Sliding Window (Hash Set):", sliding_window_set(s))

main()
